src/components/contract-boost-calculator.tsx

```diff
@@ -1,484 +1,632 @@
-	const [formState, setFormState] = useState(() => {
-		const savedFormState = typeof window === 'undefined' ? null : localStorage.getItem('formState');
-		return savedFormState ? JSON.parse(savedFormState) : formInitialState;
-	});
-
-	const [equippedArtifactsByIGN, setEquippedArtifactsByIGN] = useState([]);
-	const [equippedArtifactsListForSelectedIGN, setEquippedArtifactsListForSelectedIGN] = useState(
-		{} as any,
-	);
+const ApiUriContext = createContext<string>('missing api uri');
+
+const enum FetchState {
+	IDLE = 'idle',
+	PENDING = 'pending',
+	RETRY = 'retry',
+	SUCCESS = 'success',
+	FAILURE = 'failure',
+}
 
-	const [selectedBoostPreset, setSelectedBoostPreset] = useState('1');
-	const [availableIGNOptions, setAvailableIGNOptions] = useState([
-		{ text: 'Select an option', value: '' },
-	]);
+type CalcData = {
+	eid: string;
+	gusset?: ArtifactIntensity | undefined;
+	monocle?: ArtifactIntensity | undefined;
+	chalice?: ArtifactIntensity | undefined;
+	lifeT2: string;
+	lifeT3: string;
+	lifeT4: string;
+	diliT2: string;
+	diliT3: string;
+	diliT4: string;
+	boost: string;
+	fetchState: FetchState;
+	fetchRetryIn: number;
+	doubleDuration: boolean;
+	baseIhr: string;
+	hatcheryCalm: string;
+};
+const defaultCalcData = () => ({
+	eid: '',
+	lifeT2: '0',
+	lifeT3: '0',
+	lifeT4: '0',
+	diliT2: '0',
+	diliT3: '0',
+	diliT4: '0',
+	boost: 'boost8',
+	fetchState: FetchState.IDLE,
+	fetchRetryIn: 0,
+	doubleDuration: false,
+	baseIhr: '7440',
+	hatcheryCalm: '20',
+});
+
+const Calculator = generateCalculator<CalcData>(defaultCalcData());
+const { Input, Output } = Calculator;
+
+type FetchCoopDataProps = { readonly children: React.ReactNode };
+const FetchCoopDataButton = ({ children }: FetchCoopDataProps) => {
+	const apiUri = useContext(ApiUriContext);
+	const { data, updateData } = useContext<WithSetter<CalcData>>(Calculator.Context);
+
+	const fetchData = useCallback(async () => {
+		const backoffs = [1, 2, 5, 8, 13, 0] as const;
+		let eIResponse: EIBackupResponse | null = null;
+		for (const backoff of backoffs) {
+			updateData({ fetchState: FetchState.PENDING });
+			const rawEIResponse = await fetch(`${apiUri}/backup?EID=${data.eid}`);
+
+			if (rawEIResponse.ok) {
+				eIResponse = await rawEIResponse.json();
+				break;
+			} else if (backoff > 0) {
+				updateData({ fetchState: FetchState.RETRY, fetchRetryIn: backoff });
+				// we're doing this as a bunch of 1-second sleeps to get a countdown text
+				// this is technically not that accurate and it should be a setInterval
+				// but that was gnarly to write out
+				for (let remaining = backoff; remaining > 0; remaining--) {
+					updateData({ fetchRetryIn: remaining });
+					// no-loop-func is worried about `setTimeout` here ಠ_ಠ
+					// eslint-disable-next-line @typescript-eslint/no-loop-func
+					await new Promise((resolve) => void setTimeout(resolve, 1_000));
+				}
+			}
+		}
+
+		if (!eIResponse) {
+			updateData({ fetchState: FetchState.FAILURE });
+			return;
```

Comment:
life hack, this is an option however somewhat less descriptive
```suggestion
			return void updateData({ fetchState: FetchState.FAILURE });
```

Reply:
As a general heuristic I don't like to return `void` if my function or block is already multiple lines - so I'd def consider it if we were collapsing all the way down to
```js
if (!eiResponse) return void updateData({ fetchState: FetchState.FAILURE });
```

which... definitely isn't a bad option, now that I type it out. Thoughts?

Reply:
i support it. i usually don't open a block if the condition immediately returns or does something. i do the same with for loops
```ts
for (const x of [1, 2, 3]) console.log(x)
```